Jumpscale/data/bcdb/BCDBMeta.py

```diff
@@ -43,103 +62,114 @@ def _reset_runtime_metadata(self):
-        s = schema_obj
-        isok = j.data.schema._add_url_to_md5(s.url, s.md5)
-        if not isok:
-            # means there is a newer one in memory
-            schema = j.data.schema.get_from_url(s.url)
-            self._schema_set(schema)  # tell our bdcdb metadata that this exists
-
         # don't load the full schema but put the text of schema there
-        if not j.data.schema.exists(md5=s.md5):
-            j.data.schema._md5_to_schema[s.md5] = s.text
-
-    def _url_mid_set(self, schema):
-        url = schema.url
-        last = 0
-        if not url in self._data.url_to_mid:
-            # look for the last one
-            for item in self._data.url_to_mid.values():
-                if int(item) > last:
-                    last = int(item)
-            self._data.url_to_mid[url] = last + 1
+        if not j.data.schema.exists(md5=schema._md5):
+            j.data.schema._md5_to_schema[schema._md5] = schema.text
+
+        # need to check in j.data.schema that this schema is the newest version
+        j.shell()
+
+    def _data_from_url(self, url):
+        if url not in self._data["url"]:
+            raise j.exceptions.Input("cannot find url in bcbd:%s" % url)
+        if len(self._data["url"][url]) == 0:
+            raise j.exceptions.Input("cannot find a schema for url in bcbd:%s (no md5s)" % url)
+        md5 = self._data["url"][url][-1]
+        if md5 not in self._data["md5"]:
+            raise j.exceptions.Input("cannot find md5 in bcbd metadafta:%s" % md5)
+        return self._data["md5"][md5]
+
+    def _mid_from_url(self, url):
+        d = self._data_from_url(url)
+        return d["mid"]
 
     def _schema_exists(self, md5):
-        assert isinstance(md5, str)
-        for s in self._data.schemas:
-            # self._log_debug("check hasdata for meta:%s" % s.url)
-            if s.md5 == md5:
-                return True
-        return False
+        return md5 in self._data["md5"]
+
+    def _schema_delete(self, md5):
+        if self._schema_exists(md5):
+            self._data["md5"].pop(md5)
 
     def hasdata_set(self, schema):
         assert schema.hasdata  # needs to be True because thats what we need to set
-        for s in self._data.schemas:
-            # self._log_debug("check hasdata for meta:%s (%s:%s)" % (s.url, s.md5, schema._md5))
-            if s.md5 == schema._md5:
-                if not s.hasdata:
-                    s.hasdata = True
-                    self._save()
-                return
-        raise j.exceptions.Value("did not find schema:%s in metadata" % schema._md5)
+        if not self._schema_exists(schema._md5):
+            raise j.exceptions.Value("did not find schema:%s in metadata" % schema._md5)
+        if not self._data["md5"][schema._md5]["hasdata"]:
+            # only save when not set yet
+            self._data["md5"][schema._md5]["hasdata"] = True
+            self._save()
 
     def __repr__(self):
         return str(self._schemas_in_data_print())
```

Comment:
schemas_in_data_print doesn't return string, it `prints` and returns None

Reply:
?

Reply:
I mean the value of calling `self_schemas_in_data_print()` is `        print(j.core.tools._data_serializer_safe(self._data))` and it returns nothing to be converted to string